### DISCUSSION TITLE
Custom objective function compilation errors in reverse mode

### DISCUSSION THREAD

**Discussion by albisb** [mdolab/dafoam#498] [General]
Hi all,

I am trying to do some software development with DAFoam (e.g. defining new objective functions). My code compiles correctly in normal OpenFOAM, but when it comes to reverse mode I am getting some compilation errors, related to data types. For instance:

DAObjFunc/matrixType.C:163:33: error: cannot convert 'const codi::ActiveType<codi::JacobianLinear Tape <codi::JacobianTapeTypes<double, double, codi::Linear IndexManager<int>, codi::DefaultChunkedData, codi::LocalAdjoints>>>' to '__gnu_cxx::__alloc_traits<std::allocator<double>, double>::value_type' {aka 'double'} result[i][0] = scalars[i];

While keeping track of what data type each variable belongs to on forward OpenFOAM is straightforward, I am struggling to understand how the reverse mode is working and what I should take into account when defining the custom objective functions. Do you have a tip or indications on best practices to avoid these kind of compilation errors with ADR OpenFOAM?

Thanks

**Comment by friedenhe** [comment]
Can you paste the snippet for the problematic codes? I suspect it is caused by the variable result. Did you declare it as an OpenFOAM matrix or a List<List<scalar>>?

**Reply by albisb** [reply]
Hi, 

I am having type missmatch problems in a few parts of the code, but in the specific case I pasted, it comes from this auxiliar function I am using:
matrixx scalarListToVectormatrixx(const scalarList& scalars) {
    int numRows = scalars.size();
    matrixx result(numRows, std::vector<double>(1, 0.0));

    for (int i = 0; i < numRows; ++i) {
        result[i][0] = scalars[i];
    }

    return result;
}

As you mentioned, I have been using scalarLists. I also have defined a custom matrix type as using matrixx = std::vector<std::vector<double>>;

Do you think that may be the problem? I can try changing these to other variable types such as PETSc matrices Vec or scalarFields maybe

Thanks

**Reply by albisb** [reply]
Hi all,

I adapted the code to avoid using scalarLists for computation as well as the custom matrix type and now it goes through both in forward and reverse mode.

Thanks

**Reply by friedenhe** [reply]
Yeah, the issue is probably in your custom matrix type. The forward and reverse AD do support scalarLists.

**Reply by albisb** [reply]
Then the issue is most likely coming from the mappings I was using to transform scalarList to vector and so on. I guess a forAll loop is more robust
